core/src/main/java/com/scalar/db/transaction/consensuscommit/ParallelExecutor.java

```diff
@@ -49,79 +54,144 @@ public ParallelExecutor(ConsensusCommitConfig config) {
 
   public void prepare(List<ParallelExecutorTask> tasks) throws ExecutionException {
     try {
-      executeTasks(tasks, config.isParallelPreparationEnabled(), false);
+      executeTasks("preparation", tasks, config.isParallelPreparationEnabled(), false, true);
     } catch (CommitConflictException ignored) {
       // tasks for preparation should not throw CommitConflictException
     }
   }
 
   public void validate(List<ParallelExecutorTask> tasks)
       throws ExecutionException, CommitConflictException {
-    executeTasks(tasks, config.isParallelValidationEnabled(), false);
+    executeTasks("validation", tasks, config.isParallelValidationEnabled(), false, true);
   }
 
-  public void commit(List<ParallelExecutorTask> tasks) throws ExecutionException {
+  public void commitRecords(List<ParallelExecutorTask> tasks) throws ExecutionException {
     try {
-      executeTasks(tasks, config.isParallelCommitEnabled(), config.isAsyncCommitEnabled());
+      executeTasks(
+          "commitRecords",
+          tasks,
+          config.isParallelCommitEnabled(),
+          config.isAsyncCommitEnabled(),
+          false);
     } catch (CommitConflictException ignored) {
       // tasks for commit should not throw CommitConflictException
     }
   }
 
-  public void rollback(List<ParallelExecutorTask> tasks) throws ExecutionException {
+  public void rollbackRecords(List<ParallelExecutorTask> tasks) throws ExecutionException {
     try {
-      executeTasks(tasks, config.isParallelRollbackEnabled(), config.isAsyncRollbackEnabled());
+      executeTasks(
+          "rollbackRecords",
+          tasks,
+          config.isParallelRollbackEnabled(),
+          config.isAsyncRollbackEnabled(),
+          false);
     } catch (CommitConflictException ignored) {
       // tasks for rollback should not throw CommitConflictException
     }
   }
 
-  private void executeTasks(List<ParallelExecutorTask> tasks, boolean parallel, boolean noWait)
+  private void executeTasks(
+      String name,
+      List<ParallelExecutorTask> tasks,
+      boolean parallel,
+      boolean noWait,
+      boolean stopOnError)
       throws ExecutionException, CommitConflictException {
-    List<Future<?>> futures;
     if (parallel) {
-      assert parallelExecutorService != null;
-      futures =
-          tasks.stream()
-              .map(
-                  t ->
-                      parallelExecutorService.submit(
-                          () -> {
-                            t.run();
-                            return null;
-                          }))
-              .collect(Collectors.toList());
+      executeTasksInParallel(name, tasks, noWait, stopOnError);
     } else {
-      futures = Collections.emptyList();
-      for (ParallelExecutorTask task : tasks) {
-        task.run();
-      }
+      executeTasksSerially(name, tasks, stopOnError);
     }
+  }
+
+  private void executeTasksInParallel(
```

Comment:
With a heavy TPC-C workload, I'm facing an issue that some records remain in prepare-state even if using this branch with parallel rollback (and using the current master as well). I will check further.

Reply:
Oh okay. Thank you for checking this. 🙇